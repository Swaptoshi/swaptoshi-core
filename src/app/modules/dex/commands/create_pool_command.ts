/* eslint-disable class-methods-use-this */

import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { commandSwapContext } from '../stores/context';
import { PoolStore } from '../stores/pool';
import { createPoolCommandSchema } from '../schema/commands/create_pool_command';
import { CreatePoolParams } from '../types';
import { verifyCreatePoolParam } from '../utils/verify';
import { inversePriceSqrt } from '../utils';

export class CreatePoolCommand extends BaseCommand {
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(
		_context: CommandVerifyContext<CreatePoolParams>,
	): Promise<VerificationResult> {
		try {
			verifyCreatePoolParam(_context.params);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async execute(_context: CommandExecuteContext<CreatePoolParams>): Promise<void> {
		const poolStore = this.stores.get(PoolStore);
		const context = commandSwapContext(_context);
		const {
			tokenA,
			tokenADecimal,
			tokenASymbol,
			tokenB,
			tokenBDecimal,
			tokenBSymbol,
			fee,
			sqrtPriceX96,
		} = _context.params;
		const pool = await poolStore.createPool(
			context,
			tokenA,
			tokenASymbol,
			tokenADecimal,
			tokenB,
			tokenBSymbol,
			tokenBDecimal,
			fee,
		);
		await pool.initialize(
			tokenA.compare(tokenB) >= 0 ? inversePriceSqrt(sqrtPriceX96) : sqrtPriceX96,
		);
	}

	public schema = createPoolCommandSchema;
}
