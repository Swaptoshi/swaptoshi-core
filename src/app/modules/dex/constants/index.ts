/* eslint-disable import/no-cycle */
export * from './token_symbol';
export * from './address';
export * from './attribute';
export * from './config';
export * from './verify';
export * from './chain';
