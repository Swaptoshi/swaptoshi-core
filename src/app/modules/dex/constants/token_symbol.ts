export const DEX_DEFAULT_NAME = 'Swaptoshi';
export const DEX_DEFAULT_TOKEN_SYMBOL = 'SWX';
export const DEX_DEFAULT_TOKEN_DECIMAL = 8;

export const MAINCHAIN_DEFAULT_TOKEN_SYMBOL = 'KLY';
export const MAINCHAIN_DEFAULT_TOKEN_DECIMAL = 8;
