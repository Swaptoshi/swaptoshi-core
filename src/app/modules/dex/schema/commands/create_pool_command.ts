export const createPoolCommandSchema = {
	$id: '/dex/command/create_pool',
	type: 'object',
	properties: {
		tokenA: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		tokenASymbol: {
			dataType: 'string',
			fieldNumber: 2,
		},
		tokenADecimal: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		tokenB: {
			dataType: 'bytes',
			fieldNumber: 4,
		},
		tokenBSymbol: {
			dataType: 'string',
			fieldNumber: 5,
		},
		tokenBDecimal: {
			dataType: 'uint32',
			fieldNumber: 6,
		},
		fee: {
			dataType: 'string',
			fieldNumber: 7,
		},
		sqrtPriceX96: {
			dataType: 'string',
			fieldNumber: 8,
		},
	},
};
