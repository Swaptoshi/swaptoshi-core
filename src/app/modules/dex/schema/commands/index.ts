export * from './burn_command';
export * from './collect_command';
export * from './collect_treasury_command';
export * from './create_pool_command';
export * from './decrease_liquidity_command';
export * from './exact_input_command';
export * from './exact_input_single_command';
export * from './exact_output_command';
export * from './exact_output_single_command';
export * from './increase_liquidity_command';
export * from './mint_command';
export * from './treasurify_command';
