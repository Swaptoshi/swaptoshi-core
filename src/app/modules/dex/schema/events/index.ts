export * from './burn';
export * from './collect';
export * from './collect_position';
export * from './collect_protocol';
export * from './decrease_liquidity';
export * from './flash';
export * from './increase_liquidity';
export * from './increase_observation_cardinality_next';
export * from './mint';
export * from './pool_created';
export * from './pool_initialized';
export * from './swap';
export * from './token_registered';
export * from './tokenuri_created';
export * from './tokenuri_destroyed';
export * from './treasurify';
