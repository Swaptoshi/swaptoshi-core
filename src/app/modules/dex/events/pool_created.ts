/* eslint-disable import/no-cycle */
import { BaseEvent } from 'klayr-sdk';
import { poolCreatedEventSchema } from '../schema';

export interface PoolCreatedEventData {
	token0: Buffer;
	token1: Buffer;
	fee: string;
	tickSpacing: string;
	poolAddress: Buffer;
}

export class PoolCreatedEvent extends BaseEvent<PoolCreatedEventData> {
	public schema = poolCreatedEventSchema;
}
