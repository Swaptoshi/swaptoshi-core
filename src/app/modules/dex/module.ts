/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import {
	BaseInteroperableModule,
	BlockExecuteContext,
	FeeMethod,
	MainchainInteroperabilityMethod,
	ModuleInitArgs,
	ModuleMetadata,
	NFTMethod,
	SidechainInteroperabilityMethod,
	TokenMethod,
	TransactionExecuteContext,
	TransactionVerifyContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { BurnCommand } from './commands/burn_command';
import { CollectCommand } from './commands/collect_command';
import { CreatePoolCommand } from './commands/create_pool_command';
import { DecreaseLiquidityCommand } from './commands/decrease_liquidity_command';
import { ExactInputCommand } from './commands/exact_input_command';
import { ExactInputSingleCommand } from './commands/exact_input_single_command';
import { ExactOutputCommand } from './commands/exact_output_command';
import { ExactOutputSingleCommand } from './commands/exact_output_single_command';
import { IncreaseLiquidityCommand } from './commands/increase_liquidity_command';
import { MintCommand } from './commands/mint_command';
import { TreasurifyCommand } from './commands/treasurify_command';
import { DexEndpoint } from './endpoint';
import { BurnEvent } from './events/burn';
import { CollectEvent } from './events/collect';
import { CollectPositionEvent } from './events/collect_position';
import { CollectProtocolEvent } from './events/collect_protocol';
import { DecreaseLiquidityEvent } from './events/decrease_liquidity';
import { FlashEvent } from './events/flash';
import { IncreaseLiquidityEvent } from './events/increase_liquidity';
import { IncreaseObservationCardinalityNextEvent } from './events/increase_observation_cardinality_next';
import { MintEvent } from './events/mint';
import { PoolCreatedEvent } from './events/pool_created';
import { PoolInitializedEvent } from './events/pool_initialized';
import { SwapEvent } from './events/swap';
import { TokenURICreatedEvent } from './events/tokenuri_created';
import { TokenURIDestroyedEvent } from './events/tokenuri_destroyed';
import { DexMethod } from './method';
import {
	getPoolEndpointRequestSchema,
	getPoolEndpointResponseSchema,
} from './schema/endpoint/get_pool';
import {
	getPositionEndpointRequestSchema,
	getPositionEndpointResponseSchema,
} from './schema/endpoint/get_position';
import {
	getTokenURIEndpointRequestSchema,
	getTokenURIEndpointResponseSchema,
} from './schema/endpoint/get_token_uri';
import {
	observeEndpointRequestSchema,
	observeEndpointResponseSchema,
} from './schema/endpoint/observe';
import {
	quoteExactInputEndpointRequestSchema,
	quoteExactInputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input';
import {
	quoteExactInputSingleEndpointRequestSchema,
	quoteExactInputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_input_single';
import {
	quoteExactOutputEndpointRequestSchema,
	quoteExactOutputEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output';
import {
	quoteExactOutputSingleEndpointRequestSchema,
	quoteExactOutputSingleEndpointResponseSchema,
} from './schema/endpoint/quote_exact_output_single';
import { ObservationStore } from './stores/observation';
import { PoolStore } from './stores/pool';
import { PositionInfoStore } from './stores/position_info';
import { PositionManagerStore } from './stores/position_manager';
import { TickBitmapStore } from './stores/tick_bitmap';
import { TickInfoStore } from './stores/tick_info';
import { TokenSymbolStore } from './stores/token_symbol';
import { DexModuleConfig } from './types';
import { verifyMinimumFee, verifySwapByTransfer, executeSwapByTransfer } from './hooks';
import { defaultConfig } from './constants';
import {
	getMetadataEndpointRequestSchema,
	getMetadataEndpointResponseSchema,
} from './schema/endpoint/get_metadata';
import { verifyFeeConversion } from './hooks/verifyFeeConversion';
import { executeFeeConversion } from './hooks/executeFeeConversion';
import { TreasurifyEvent } from './events/treasurify';
import {
	getPoolAddressFromCollectionIdEndpointRequestSchema,
	getPoolAddressFromCollectionIdEndpointResponseSchema,
} from './schema/endpoint/get_pool_address_from_collection_id';
import {
	quotePriceEndpointRequestSchema,
	quotePriceEndpointResponseSchema,
} from './schema/endpoint/quote_price';
import { TokenRegisteredEvent } from './events/token_registered';
import { SupportedTokenStore } from './stores/supported_token';
import { verifyValidTransfer } from './hooks/verifyValidTransfer';
import { DexInteroperableMethod } from './cc_method';

export class DexModule extends BaseInteroperableModule {
	public _config: DexModuleConfig | undefined;
	public _feeMethod: FeeMethod | undefined;
	public _tokenMethod: TokenMethod | undefined;
	public _dexInteroperableMethod = new DexInteroperableMethod(this.stores, this.events);

	public crossChainCommand = [];
	public crossChainMethod = this._dexInteroperableMethod;

	public endpoint = new DexEndpoint(this.stores, this.offchainStores);
	public method = new DexMethod(this.stores, this.events);
	public commands = [
		new CreatePoolCommand(this.stores, this.events),
		new MintCommand(this.stores, this.events),
		new IncreaseLiquidityCommand(this.stores, this.events),
		new DecreaseLiquidityCommand(this.stores, this.events),
		new CollectCommand(this.stores, this.events),
		new BurnCommand(this.stores, this.events),
		new ExactInputCommand(this.stores, this.events),
		new ExactInputSingleCommand(this.stores, this.events),
		new ExactOutputCommand(this.stores, this.events),
		new ExactOutputSingleCommand(this.stores, this.events),
		new TreasurifyCommand(this.stores, this.events),
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(PoolStore, new PoolStore(this.name, 0, this.stores, this.events));
		this.stores.register(
			PositionManagerStore,
			new PositionManagerStore(this.name, 1, this.stores, this.events),
		);

		this.stores.register(ObservationStore, new ObservationStore(this.name, 2));
		this.stores.register(PositionInfoStore, new PositionInfoStore(this.name, 3));
		this.stores.register(TickBitmapStore, new TickBitmapStore(this.name, 4));
		this.stores.register(TickInfoStore, new TickInfoStore(this.name, 5));
		this.stores.register(TokenSymbolStore, new TokenSymbolStore(this.name, 6, this.events));
		this.stores.register(SupportedTokenStore, new SupportedTokenStore(this.name, 7));

		this.events.register(BurnEvent, new BurnEvent(this.name));
		this.events.register(CollectPositionEvent, new CollectPositionEvent(this.name));
		this.events.register(CollectProtocolEvent, new CollectProtocolEvent(this.name));
		this.events.register(CollectEvent, new CollectEvent(this.name));
		this.events.register(DecreaseLiquidityEvent, new DecreaseLiquidityEvent(this.name));
		this.events.register(FlashEvent, new FlashEvent(this.name));
		this.events.register(IncreaseLiquidityEvent, new IncreaseLiquidityEvent(this.name));
		this.events.register(
			IncreaseObservationCardinalityNextEvent,
			new IncreaseObservationCardinalityNextEvent(this.name),
		);
		this.events.register(MintEvent, new MintEvent(this.name));
		this.events.register(PoolCreatedEvent, new PoolCreatedEvent(this.name));
		this.events.register(PoolInitializedEvent, new PoolInitializedEvent(this.name));
		this.events.register(SwapEvent, new SwapEvent(this.name));
		this.events.register(TokenURICreatedEvent, new TokenURICreatedEvent(this.name));
		this.events.register(TokenURIDestroyedEvent, new TokenURIDestroyedEvent(this.name));
		this.events.register(TreasurifyEvent, new TreasurifyEvent(this.name));
		this.events.register(TokenRegisteredEvent, new TokenRegisteredEvent(this.name));
	}

	public addDependencies(
		tokenMethod: TokenMethod,
		nftMethod: NFTMethod,
		feeMethod: FeeMethod,
		interoperabilityMethod: SidechainInteroperabilityMethod | MainchainInteroperabilityMethod,
	) {
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportManagerStore = this.stores.get(SupportedTokenStore);

		poolStore.addDependencies(tokenMethod);
		positionManagerStore.addDependencies(tokenMethod, nftMethod);
		supportManagerStore.addDependencies(tokenMethod);

		this._feeMethod = feeMethod;
		this._tokenMethod = tokenMethod;
		this._dexInteroperableMethod.addDependencies(interoperabilityMethod, tokenMethod, nftMethod);
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				{
					name: this.endpoint.getPool.name,
					request: getPoolEndpointRequestSchema,
					response: getPoolEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPosition.name,
					request: getPositionEndpointRequestSchema,
					response: getPositionEndpointResponseSchema,
				},
				{
					name: this.endpoint.getTokenURI.name,
					request: getTokenURIEndpointRequestSchema,
					response: getTokenURIEndpointResponseSchema,
				},
				{
					name: this.endpoint.getPoolAddressFromCollectionId.name,
					request: getPoolAddressFromCollectionIdEndpointRequestSchema,
					response: getPoolAddressFromCollectionIdEndpointResponseSchema,
				},
				{
					name: this.endpoint.getMetadata.name,
					request: getMetadataEndpointRequestSchema,
					response: getMetadataEndpointResponseSchema,
				},
				{
					name: this.endpoint.observe.name,
					request: observeEndpointRequestSchema,
					response: observeEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInput.name,
					request: quoteExactInputEndpointRequestSchema,
					response: quoteExactInputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactInputSingle.name,
					request: quoteExactInputSingleEndpointRequestSchema,
					response: quoteExactInputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutput.name,
					request: quoteExactOutputEndpointRequestSchema,
					response: quoteExactOutputEndpointResponseSchema,
				},
				{
					name: this.endpoint.quoteExactOutputSingle.name,
					request: quoteExactOutputSingleEndpointRequestSchema,
					response: quoteExactOutputSingleEndpointResponseSchema,
				},
				{
					name: this.endpoint.quotePrice.name,
					request: quotePriceEndpointRequestSchema,
					response: quotePriceEndpointResponseSchema,
				},
			],
			assets: [],
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(_args: ModuleInitArgs): Promise<void> {
		this._config = { ...defaultConfig, ..._args.moduleConfig };
		const poolStore = this.stores.get(PoolStore);
		const positionManagerStore = this.stores.get(PositionManagerStore);
		const supportedTokenStore = this.stores.get(SupportedTokenStore);

		poolStore.init(this._config);
		positionManagerStore.init(Buffer.from(_args.genesisConfig.chainID, 'hex'));
		supportedTokenStore.init(this._config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
		try {
			await verifyMinimumFee(_context, this._config!);
			await verifyValidTransfer.bind(this)(_context);
			await verifySwapByTransfer.bind(this)(_context);
			await verifyFeeConversion.bind(this)(_context);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await verifyMinimumFee(_context, this._config!);
		await verifyValidTransfer.bind(this)(_context);
		await verifySwapByTransfer.bind(this)(_context);
		await executeFeeConversion.bind(this)(_context);
	}

	public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
		await executeSwapByTransfer.bind(this)(_context);
	}

	public async beforeTransactionsExecute(context: BlockExecuteContext): Promise<void> {
		await this.stores.get(SupportedTokenStore).apply(context);
	}
}
