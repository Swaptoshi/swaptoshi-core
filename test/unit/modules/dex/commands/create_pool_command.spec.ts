/* eslint-disable camelcase */
/* eslint-disable jest/expect-expect */
import { StateMachine } from 'klayr-sdk';
import { CreatePoolCommand } from '../../../../../src/app/modules/dex/commands/create_pool_command';
import { DexModule } from '../../../../../src/app/modules/dex/module';
import { createPoolCommandSchema } from '../../../../../src/app/modules/dex/schema';
import { CreatePoolParams, MutableSwapContext, TokenMethod } from '../../../../../src/app/modules/dex/types';
import { invalidNumberString, invalidTokenAddress } from '../utils/invalid';
import { commandFixture } from '../utils/fixtures';
import { FeeAmount, encodePriceSqrt } from '../stores/shared/utilities';
import { poolAddress2, senderPublicKey, token0, token1, token2 } from '../utils/account';
import { eventResultHaveMinimumLength } from '../../../../utils/events';
import { NFTRegistry } from '../stores/shared/nft/nft_registry';
import { TokenRegistry } from '../stores/shared/token/token_registry';
import { PoolCreatedEvent } from '../../../../../src/app/modules/dex/events/pool_created';
import { PoolStore } from '../../../../../src/app/modules/dex/stores/pool';
import { commandSwapContext } from '../../../../../src/app/modules/dex/stores/context';
import { mock_token_initializeUserAccount, mock_token_lock, mock_token_transfer } from '../stores/shared/token';
import { DEFAULT_TREASURY_ADDRESS, POSITION_MANAGER_ADDRESS, ROUTER_ADDRESS, defaultConfig } from '../../../../../src/app/modules/dex/constants';
import { PositionManagerStore } from '../../../../../src/app/modules/dex/stores/position_manager';
import { TokenSymbolStore } from '../../../../../src/app/modules/dex/stores/token_symbol';

type CommandParam = CreatePoolParams;
const COMMAND_NAME = 'createPool';
const commandSchema = createPoolCommandSchema;

const validParam: CommandParam = {
	tokenA: token1,
	tokenASymbol: 'TKNA',
	tokenADecimal: 8,
	tokenB: token2,
	tokenBSymbol: 'TKNB',
	tokenBDecimal: 8,
	fee: FeeAmount.MEDIUM,
	sqrtPriceX96: encodePriceSqrt(1, 1).toString(),
};

describe('CreatePoolCommand', () => {
	let module: DexModule;
	let command: CreatePoolCommand;
	let swapContext: MutableSwapContext;
	let poolStore: PoolStore;
	let positionManagerStore: PositionManagerStore;
	let tokenSymbolStore: TokenSymbolStore;
	let tokenMethod: TokenMethod;
	let createCommandVerifyContext: (params: CommandParam) => StateMachine.CommandVerifyContext<CommandParam>;
	let createCommandExecuteContext: (params: CommandParam) => StateMachine.CommandExecuteContext<CommandParam>;

	beforeEach(async () => {
		({ module, createCommandExecuteContext, createCommandVerifyContext, poolStore, tokenMethod, positionManagerStore, tokenSymbolStore } = await commandFixture<CommandParam>(
			COMMAND_NAME,
			commandSchema,
			senderPublicKey,
			validParam,
		));
		command = new CreatePoolCommand(module.stores, module.events);
	});

	afterEach(() => {
		NFTRegistry.reset();
		TokenRegistry.reset();
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toBe(COMMAND_NAME);
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		it('should return status OK when called with valid input', async () => {
			const context = createCommandVerifyContext(validParam);
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.OK);
		});

		it('should throw error when user sends transaction with invalid token address (tokenA)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				tokenA: invalidTokenAddress,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid token address (tokenB)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				tokenB: invalidTokenAddress,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (fee)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				fee: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});

		it('should throw error when user sends transaction with invalid number string (sqrtPriceX96)', async () => {
			const context = createCommandVerifyContext({
				...validParam,
				sqrtPriceX96: invalidNumberString,
			});
			await expect(command.verify(context)).resolves.toHaveProperty('status', StateMachine.VerifyStatus.FAIL);
		});
	});

	describe('execute', () => {
		it('should create pool', async () => {
			const context = createCommandExecuteContext(validParam);
			swapContext = commandSwapContext(context);
			await command.execute(context);

			const pool = await poolStore.getMutablePool(swapContext, token1, token2, FeeAmount.MEDIUM);
			expect(pool.address).toStrictEqual(poolAddress2);
		});

		it('should create position manager', async () => {
			const context = createCommandExecuteContext(validParam);
			swapContext = commandSwapContext(context);
			await command.execute(context);

			const positionManager = await positionManagerStore.getMutablePositionManager(swapContext, poolAddress2);
			expect(positionManager.poolAddress).toStrictEqual(poolAddress2);
		});

		it('should register token symbol and decimals', async () => {
			const context = createCommandExecuteContext(validParam);
			swapContext = commandSwapContext(context);
			await command.execute(context);

			const token1Symbol = await tokenSymbolStore.has(context, tokenSymbolStore.getKey(token1));
			const token2Symbol = await tokenSymbolStore.has(context, tokenSymbolStore.getKey(token2));
			expect(token1Symbol).toBe(true);
			expect(token2Symbol).toBe(true);
		});

		it('should throw an error if created with unsupported tickSpacing', async () => {
			const func = async () => {
				const context = createCommandExecuteContext({
					...validParam,
					fee: '25',
				});
				await command.execute(context);
			};
			await expect(func()).rejects.toThrow('tickSpacing unsupported');
		});

		it('should throw an error if pool already exists', async () => {
			const func = async () => {
				const context = createCommandExecuteContext({
					...validParam,
					tokenA: token0,
					tokenB: token1,
					fee: FeeAmount.MEDIUM,
				});
				await command.execute(context);
			};
			await expect(func()).rejects.toThrow('pool already exists');
		});

		it('should initialize neccessary account', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(poolAddress2, token1);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(poolAddress2, token2);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(POSITION_MANAGER_ADDRESS, token1);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(POSITION_MANAGER_ADDRESS, token2);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(ROUTER_ADDRESS, token1);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(ROUTER_ADDRESS, token2);
		});

		it('should initialize treasury account if avaialble in config', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(DEFAULT_TREASURY_ADDRESS, token1);

			expect(mock_token_initializeUserAccount).toHaveBeenCalledWith(DEFAULT_TREASURY_ADDRESS, token2);
		});

		it('should transfer leftover balance to treasury', async () => {
			const context = createCommandExecuteContext(validParam);

			await tokenMethod.mint(context, poolAddress2, token1, BigInt(10));
			await tokenMethod.mint(context, poolAddress2, token2, BigInt(10));

			await command.execute(context);

			expect(mock_token_transfer).toHaveBeenCalledWith(poolAddress2, DEFAULT_TREASURY_ADDRESS, token1, BigInt(10));

			expect(mock_token_transfer).toHaveBeenCalledWith(poolAddress2, DEFAULT_TREASURY_ADDRESS, token2, BigInt(10));
		});

		it('should lock leftover balance if treasury is not configured', async () => {
			module._config.default = {
				...defaultConfig,
				feeProtocolPool: '',
			};
			poolStore.init(module._config);

			const context = createCommandExecuteContext(validParam);

			await tokenMethod.mint(context, poolAddress2, token1, BigInt(10));
			await tokenMethod.mint(context, poolAddress2, token2, BigInt(10));

			await command.execute(context);

			expect(mock_token_lock).toHaveBeenCalledWith(poolAddress2, module.name, token1, BigInt(10));

			expect(mock_token_lock).toHaveBeenCalledWith(poolAddress2, module.name, token2, BigInt(10));
		});

		it('should add command events', async () => {
			const context = createCommandExecuteContext(validParam);
			await command.execute(context);

			eventResultHaveMinimumLength(context.eventQueue, PoolCreatedEvent, module.name, 1);
		});
	});
});
