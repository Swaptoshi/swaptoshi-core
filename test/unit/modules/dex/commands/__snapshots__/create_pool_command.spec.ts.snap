// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePoolCommand constructor should have valid schema 1`] = `
{
  "$id": "/dex/command/create_pool",
  "properties": {
    "fee": {
      "dataType": "string",
      "fieldNumber": 7,
    },
    "sqrtPriceX96": {
      "dataType": "string",
      "fieldNumber": 8,
    },
    "tokenA": {
      "dataType": "bytes",
      "fieldNumber": 1,
      "format": "hex",
    },
    "tokenADecimal": {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "tokenASymbol": {
      "dataType": "string",
      "fieldNumber": 2,
    },
    "tokenB": {
      "dataType": "bytes",
      "fieldNumber": 4,
      "format": "hex",
    },
    "tokenBDecimal": {
      "dataType": "uint32",
      "fieldNumber": 6,
    },
    "tokenBSymbol": {
      "dataType": "string",
      "fieldNumber": 5,
    },
  },
  "type": "object",
}
`;
